File: patterns/concurrency/read-write-lock.js
/**
 * Read-Write Lock Pattern - JavaScript
 * 
 * Allows multiple readers or a single writer to access a resource,
 * improving concurrency for read-heavy workloads.
 */

class ReadWriteLock {
    constructor() {
        this.readers = 0;
        this.writers = 0;
        this.waitingWriters = 0;
        this.lock = false;
    }
    
    async acquireReadLock() {
        while (this.writers > 0 || this.waitingWriters > 0) {
            await new Promise(resolve => setTimeout(resolve, 10));
        }
        this.readers++;
    }
    
    releaseReadLock() {
        this.readers--;
    }
    
    async acquireWriteLock() {
        this.waitingWriters++;
        while (this.readers > 0 || this.writers > 0) {
            await new Promise(resolve => setTimeout(resolve, 10));
        }
        this.waitingWriters--;
        this.writers++;
    }
    
    releaseWriteLock() {
        this.writers--;
    }
}

class SharedResource {
    constructor() {
        this.data = 0;
        this.lock = new ReadWriteLock();
    }
    
    async read() {
        await this.lock.acquireReadLock();
        const value = this.data;
        this.lock.releaseReadLock();
        return value;
    }
    
    async write(value) {
        await this.lock.acquireWriteLock();
        this.data = value;
        this.lock.releaseWriteLock();
    }
}

// Example usage
console.log('=== Read-Write Lock Pattern ===\n');

const resource = new SharedResource();

async function reader(id) {
    const value = await resource.read();
    console.log(`Reader ${id} read: ${value}`);
}

async function writer(id, value) {
    await resource.write(value);
    console.log(`Writer ${id} wrote: ${value}`);
}

(async () => {
    await Promise.all([
        reader(1),
        reader(2),
        writer(1, 10),
        reader(3),
        writer(2, 20),
        reader(4)
    ]);
})();

module.exports = { ReadWriteLock, SharedResource };

File: patterns/concurrency/read-write-lock.py
"""
Read-Write Lock Pattern - Python

Allows multiple readers or a single writer to access a resource,
improving concurrency for read-heavy workloads.
"""

import threading
import time


class ReadWriteLock:
    def __init__(self):
        self._read_ready = threading.Condition(threading.RLock())
        self._readers = 0
    
    def acquire_read(self):
        self._read_ready.acquire()
        try:
            self._readers += 1
        finally:
            self._read_ready.release()
    
    def release_read(self):
        self._read_ready.acquire()
        try:
            self._readers -= 1
            if self._readers == 0:
                self._read_ready.notifyAll()
        finally:
            self._read_ready.release()
    
    def acquire_write(self):
        self._read_ready.acquire()
        while self._readers > 0:
            self._read_ready.wait()
    
    def release_write(self):
        self._read_ready.release()


class SharedResource:
    def __init__(self):
        self.data = 0
        self.lock = ReadWriteLock()
    
    def read(self):
        self.lock.acquire_read()
        try:
            return self.data
        finally:
            self.lock.release_read()
    
    def write(self, value):
        self.lock.acquire_write()
        try:
            self.data = value
        finally:
            self.lock.release_write()


# Example usage
if __name__ == "__main__":
    print("=== Read-Write Lock Pattern ===\n")
    
    resource = SharedResource()
    
    def reader(id):
        value = resource.read()
        print(f"Reader {id} read: {value}")
    
    def writer(id, value):
        resource.write(value)
        print(f"Writer {id} wrote: {value}")
    
    threads = [
        threading.Thread(target=reader, args=(1,)),
        threading.Thread(target=reader, args=(2,)),
        threading.Thread(target=writer, args=(1, 10)),
        threading.Thread(target=reader, args=(3,)),
        threading.Thread(target=writer, args=(2, 20)),
        threading.Thread(target=reader, args=(4,))
    ]
    
    for t in threads:
        t.start()
    
    for t in threads:
        t.join()

